services/src/datasets/external/sentinel_s2_l2a_cogs.rs

```diff
@@ -0,0 +1,640 @@
+
+        for page in 2..=num_pages {
+            let mut collection = self.load_collection(params, page).await?;
+            features.append(&mut collection.features);
+        }
+
+        Ok(features)
+    }
+
+    async fn load_collection<T: Serialize + ?Sized>(
+        &self,
+        params: &T,
+        page: u32,
+    ) -> Result<StacCollection> {
+        let client = reqwest::Client::new();
+        let text = client
+            .get(&self.api_url)
+            .query(&params)
+            .query(&[("page", &page.to_string())])
+            .send()
+            .await
+            .context(error::Reqwest)?
+            .text()
+            .await
+            .context(error::Reqwest)?;
+
+        serde_json::from_str(&text).map_err(|error| error::Error::StacJsonRespone {
+            url: self.api_url.clone(),
+            response: text,
+            error,
+        })
+    }
+
+    fn time_range_request(time: &TimeInterval) -> Result<(DateTime<Utc>, DateTime<Utc>)> {
+        let t_start =
+            time.start()
+                .as_utc_date_time()
+                .ok_or(geoengine_operators::error::Error::DataType {
+                    source: geoengine_datatypes::error::Error::NoDateTimeValid {
+                        time_instance: time.start(),
+                    },
+                })?;
+
+        // shift start by 1 day to ensure getting the most recent data for start time
+        let t_start = t_start - Duration::days(1);
+
+        let t_end =
+            time.end()
+                .as_utc_date_time()
+                .ok_or(geoengine_operators::error::Error::DataType {
+                    source: geoengine_datatypes::error::Error::NoDateTimeValid {
+                        time_instance: time.end(),
+                    },
+                })?;
+
+        Ok((t_start, t_end))
+    }
+}
+
+#[async_trait]
+impl MetaData<GdalLoadingInfo, RasterResultDescriptor> for SentinelS2L2aCogsMetaData {
+    async fn loading_info(
+        &self,
+        query: QueryRectangle,
+    ) -> geoengine_operators::util::Result<GdalLoadingInfo> {
+        // TODO: propagate error properly
+        self.create_loading_info(query).await.map_err(|e| {
+            geoengine_operators::error::Error::LoadingInfo {
+                reason: format!("{:?}", e),
+            }
+        })
+    }
+
+    async fn result_descriptor(&self) -> geoengine_operators::util::Result<RasterResultDescriptor> {
+        Ok(RasterResultDescriptor {
+            data_type: self.band.data_type,
+            spatial_reference: SpatialReference::new(
+                SpatialReferenceAuthority::Epsg,
+                self.zone.epsg,
+            )
+            .into(),
+            measurement: Measurement::Unitless,
+            no_data_value: self.band.no_data_value,
+        })
+    }
+
+    fn box_clone(&self) -> Box<dyn MetaData<GdalLoadingInfo, RasterResultDescriptor>> {
```

Comment:
can we default impl this?

Reply:
I guess we could if we create another trait CloneableMetaData that has only that function. Shall I try it?

Reply:
no then it is to much overhead. cant you simply implement it in the MetaData trait?

Reply:
I don't know how. I used explicitly this workaround of having a box_clone method and implementing Clone for the Box because it's not possible IMHO to directly make the trait Clone

Reply:
then we have to keep it like this